### PR TITLE
Weather embedding gpu join

### PR DESCRIPTION
**Pull request by tsgra**
This code removes the weather features from the batches of training data and instead adds in an index corresponding to a weather file city. The weather features themselves are then joined via an embedding layer to the training data based on the index, and the weather data proceeds to the convolutional neural network. 

Hard for me to test whether this had the desired effect of reducing training time without running the full training pipeline. I was also not able to test training with GPUs. Perhaps @vengroff has a way? Also, question probably for @mikivee , is the `src/model_training.py` the same notebook used in the job that trains the model? The results after only 2 epochs and based on only 1000 rows of training data yielded some nonsensical results when predicting, hopefully that is by design and the non-job version of the notebook is just used to test whether the full training pipeline works.

### CODE REVIEW

File: src/datagen.py [hunk 1]
```diff
@@ -231,15 +232,15 @@ def get_weather_feature_lookups(self) -> FeatureLookup:
     def init_training_set(
         self,
         train_data: DataFrame,
-        exclude_columns: List[str] = ["building_id", "upgrade_id", "weather_file_city"],
+        exclude_columns: List[str] = ["building_id", "upgrade_id"],
```
Comment (tsgra) [inline src/datagen.py:235]: Need the weather file city here to merge to the `weather_features_df` and extract the `weather_file_city_index` for prediction in the `src/model_training.py` notebook.

File: src/model_training.py [hunk 2]
```diff
@@ -312,8 +312,13 @@ def convert_feature_dataframe_to_dict(
 test_gen = DataGenerator(train_data=test_data.limit(10))
 # load input data table as a Spark DataFrame
 input_data = test_gen.training_set.load_df().toPandas()
+# Merge to weather file city index 
+input_data_weather_file_city_index = (
+    input_data
+    .merge(test_gen.weather_features_df[['weather_file_city', 'weather_file_city_index']], on = 'weather_file_city', how = 'left')
```
Comment (tsgra) [inline src/model_training.py:318]: This is a bit janky - ideally we want the `weather_file_city_index` to be included in the `test_gen.training_set` DF, but there isn't an easy way to do this such that the city <> index is the same as the `weather_features_df`

File: src/model_training.py [hunk 3]
```diff
@@ -312,8 +312,13 @@ def convert_feature_dataframe_to_dict(
...
+)
 # run prediction and output a N x M matrix of predictions where N is the number of rows in the input data table and M is the number of target columns
-print(model_loaded.predict(input_data))
+print(model_loaded.predict(input_data_weather_file_city_index))
```
Comment (tsgra) [inline src/model_training.py:321]: This is where I was seeing nonseniscal results

**Comment by vengroff** [comment]
The key to seeing whether the change had the desired effect on GPU performance is to run a full production run, but kill it after about 5-10 epochs. Meanwhile, run the GPU tracker that I created in another tab to get a chart of how the GPU is being used.

To answer your other question, it is totally expected that results on the debug data are nonsensical. There has been nowhere near enough data run through the training process to expect the network to have evolved very far from its initial random state.

File: src/model_training.py (hunk 2 above)
Comment (vengroff) [reply src/model_training.py:318]: The most import key to all of this, however it is done, is that once an index has been assigned to a weather city it can never change for the lifetime of the model, no matter how or where inference is done.

File: src/model_training.py (hunk 3 above)
Comment (vengroff) [reply src/model_training.py:321]: If you have only run a tiny amount of training data through the fitting than you can't expect reasonable results yet. Have you run a single non-debug run? If so, then you should start to see convergence in the mlflow charts within an hour or so and it should be comparable to the experiments done before this branch was creates. But I think you said you have not been able to run on a GPU. We need to fix that before you can try this.

**Comment by vengroff** [comment]
I just fired up a [GPU cluster](https://4617764665359845.5.gcp.databricks.com/compute/clusters/0814-134012-bussdsk9?o=4617764665359845) so I can give this a shot. I'll run `model_training` in prod mode with monitoring and let it go for half an hour or so and see what the GPU utilization curves and mlflow curves look like. You should be able to clone that cluster and do the same as needed to study and debug this, and then when it is loking good we should run a full run to completion.

File: src/model_training.py (hunk 2 above)
Comment (mikivee) [reply src/model_training.py:318]: Also I think that as is, this join to grab the weather city index would always have to be applied at inference time. This means we would need to apply this fix somewhere [here](https://github.com/rewiringamerica/surrogate_modeling/blob/662bb2eb4a96fd1d7b8724e3236bbe06b5da7416/src/model_evaluation.py#L90) too, right? Not to mention by the caller at inference time inside dohyo as well, which does not seem ideal. 

As a result of the additional code needed in `model_evaluation.py`, I'm gonna consider this a blocker. I think the suggestion I made will address this issue tho.

File: src/surrogate_model.py [hunk 4]
```diff
@@ -129,22 +129,32 @@ def conv_batchnorm_relu(x:tf.keras.layers, filters:int, kernel_size:int, name:st
...
+                    output_dim=num_hours,
+                    weights=[np.array([city[i] for city in train_gen.weather_features_matrix])],
+                    trainable=False, 
+                    name=f'{feature_name}_embedding'
+                )(weather_file_city_index_input)
             )
-            for weather_feature in train_gen.weather_features
-        }
-        weather_inputs = list(weather_inputs_dict.values())
+            for i, feature_name in zip(range(len(train_gen.weather_features)), train_gen.weather_features)
```
Comment (mikivee) [inline src/surrogate_model.py:151]: NIT: wouldn't enumerate wok here?

**Review by mikivee — CHANGES_REQUESTED** [review]
Awesome job on this Talor!! Really excited about these improvements and you learned this code base, not to mention the basics of NNs and CNNs, incredibly quickly. 

I have [one issue with the implementation](https://github.com/rewiringamerica/surrogate_modeling/pull/17#discussion_r1750449670) that I think we should try and simplify before merging since it complicates calling at inference time. Happy to discuss offline if that would be easier.

File: src/build_feature_store.py [hunk 5]
```diff
@@ -1268,20 +1268,14 @@ def transform_weather_features() -> DataFrame:
...
             if c not in weather_pkeys + ["datetime_formatted"]
         ]
     )
-    return weather_data_arrays
-    weather_data_arrays = weather_df.groupBy(weather_pkeys).agg(
-        *[
-            F.collect_list(c).alias(c)
-            for c in weather_df.columns
-            if c not in weather_pkeys + ["datetime_formatted"]
-        ]
```
[the remark this answers is not in the document]
Comment (tsgra) [reply src/build_feature_store.py:1284]: Correct!

File: src/model_training.py (hunk 2 above)
Comment (tsgra) [reply src/model_training.py]: I've added the `weather_file_index` to the `build_feature_store.py` so it will be included in all training sets and doesn't have to be subsequently merged at the batch creation.

File: src/surrogate_model.py (hunk 4 above)
Comment (tsgra) [reply src/surrogate_model.py]: Good call, added

File: src/datagen.py [hunk 6]
```diff
@@ -196,6 +196,7 @@ def __init__(
         )
 
         self.weather_features_df = self.init_weather_features()
+        self.weather_features_matrix = self.weather_features_df.sort_values(by = 'weather_file_city')[self.weather_features].to_numpy()
```
Comment (tsgra) [inline src/datagen.py:199]: Sorting values alphabetically by the weather file city should ensure that the row numbers correspond to the same index output by the `StringIndexer` which assigns the indices in alphabetical order

Comment (vengroff) [reply src/datagen.py:197]: I think we have to be very careful about sorting in two different places and hoping for the best. I would be much happier if there were a single persisted mapping between city names and indices that was used at both training and inference time and tied to a fitted model version. If you mess up the mapping for any reason then you start serving garbage predictions and there is absolutely nothing to indicate that this is happening. You just used weather index 47 instead of 137 and got the wrong answer. I'd really rather not have any "should" in the approach we take to this.

**Review by mikivee — COMMENTED** [review]
Looks mostly good--thanks for implementing my suggestion.  still a few issues I think.

File: src/datagen.py (hunk 6 above)
Comment (vengroff) [reply src/datagen.py:197]: OK, I think that makes sense, @mikivee. I guess the final step is that when we serve this to PEP or the world at large we have to make sure that we load the exact right version of the two columns for the weather city and corresponding index and when the caller gives us an address, we geocode to a weather city, look up the index, and pass that to the model for inference.

File: src/model_training.py [hunk 7]
```diff
@@ -214,7 +214,7 @@ def convert_feature_dataframe_to_dict(
         X_train_bm = {col: np.array(feature_df[col]) for col in self.building_features}
         X_train_weather = {
             col: np.array(np.vstack(feature_df[col].values))
-            for col in self.weather_features
+            for col in ['weather_file_city_index']
         }
         return {**X_train_bm, **X_train_weather}
```
[the remark this answers is not in the document]
Comment (tsgra) [reply src/model_training.py:219]: This looks good, i'll add.

File: src/datagen.py
[the remark this answers is not in the document]
Comment (tsgra) [reply src/datagen.py:1]: In the current datagen i'm looking at it seems to not reflect the changes in that commit.

[the remark this answers is not in the document]
Comment (mikivee) [reply src/datagen.py:1]: hmm i don't really understand what could explain the discrepancy in our views but I am still seeing the following in that section of the `building_features` list:

```
  # attached home
  "is_attached",
  "n_building_units",
  "is_middle_unit"
   # other appliances
```
where in main it is: 
```
# building type
"is_attached",
"is_mobile_home",
"n_building_units",
"is_middle_unit",
"unit_level_in_building",
# other appliances
```
So i think this still needs to be fixed?

[the remark this answers is not in the document]
Comment (tsgra) [reply src/datagen.py:1]: In all of my files (on databricks and locally) it looks like the `main` version.

File: src/model_training.py
Comment (mikivee) [inline src/model_training.py:1]: boo still getting the retracing issue so I guess that didn't fully fix it.. but thats cool, definitely can just add this to the list to deal w in a later PR.

File: src/model_training.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikivee) [reply src/model_training.py:219]: I'm getting the following output from the inference test at the bottom:
```
[[  0.      155.4103        nan       nan]
 [  0.      231.60963       nan       nan]
 [  0.            nan       nan 878.4548 ]
 [  0.            nan       nan       nan]
 [  0.      197.33766       nan       nan]
 [  0.      157.5082        nan       nan]
 [  0.      220.64223       nan       nan]
 [  0.      138.9264        nan       nan]
 [  0.      217.9236        nan       nan]
 [  0.            nan       nan       nan]]
```
I'm guessing the nans are cause of this vstack remnant.

File: src/datagen.py
[the remark this answers is not in the document]
Comment (mikivee) [reply src/datagen.py:1]: discussed offline-- turns out was I having a db <> gh sync issue. resolving!

**Review by mikivee — COMMENTED** [review]
Ugh sorry talor still a minor blocking issue, but should be relatively quick. Otherwise this looks great and much better!! Thank you!

Should we also run a similar test like what darren did to test for gpu utilization and speed w this final version, as well as a full test run with the eval metrics to make sure everything is working as expected? I feel like it might be good to do that before merging.

File: src/surrogate_model.py [hunk 8]
```diff
@@ -134,26 +134,20 @@ def conv_batchnorm_relu(x:tf.keras.layers, filters:int, kernel_size:int, name:st
         # Extract number of cities and hours from the weather data for dimensions of embedding layers
         num_cities = len(train_gen.weather_features_matrix)
         num_hours = len(train_gen.weather_features_matrix[0][0])
+        num_features = len(train_gen.weather_features_matrix[0])
```
[the remark this answers is not in the document]
Comment (tsgra) [reply src/surrogate_model.py:137]: Sure that looks good

File: src/model_training.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tsgra) [reply src/model_training.py:219]: Ah true.

[the remark this answers is not in the document]
Comment (tsgra) [reply src/model_training.py]: Still get a bunch of `nan` after updating the v-stack

[the remark this answers is not in the document]
Comment (tsgra) [reply src/model_training.py]: I think it's because only 2 epochs and 1000 samples

[the remark this answers is not in the document]
Comment (mikivee) [reply src/model_training.py]: wow that was dumb of me-- i literally wrote that fn and I forgot. nan just means that there is none of that fuel type in the house. my bad! regardless i think good to fix the vstack thing-- im honestly surprised it didn't error out.

**Review by mikivee — APPROVED** [review]
thanks for the final fixes! looks good to me! hopefully all the accuracy metrics still look good! 🤞

**Review by mikivee — APPROVED** [review]
Solution makes sense to me! thanks for finding
